### PR TITLE
feat: Messages v2

### PR DESCRIPTION
**Pull request by NunoAlexandre**
`WIP`

### Changes
- Update and add all v2 messages to the _latest_ Connectors v2 spec
- Implement and test flows for incoming and outgoing messages

### To Do

- [x] Merge the latest `main` (fix merge conflicts 😈 )

### Callouts

- [x] Can we avoid the need for two mappings for the currencies handling? 
  We now have one `address => uint128` and the reverse, `uint128 => address`. The latter is only used in `allowPoolCurrency` to verify that the currency being allowed for a given pool exists in the first place;
  -> We can choose to pass `address` instead of `uint128` in `AllowPoolCurrency` but that opens the question to why using `uint128` at all, specially if one same currency can't be used in different EVM domains?

-  [ ] There's a lot of inconsistent naming right now, mostly my fault 🙊  For example, some fields are called `recipient`, others `receiver`, others `destinationAddress` (albeit there being some logical different between some of these). I would like to agree on one and make that consistent across the repo.

- [ ] For the new outgoing messages I have introduced a slightly different test setup where we test all the edge cases in the same test instead of repeating a lot of boilerplate setup code just to test that the function fails if some call wasn't executed beforehand (e.g. `CollectInvest` being called by a `msg.sender` before we call `updateMember` with said user). Would like to hear your views on that (@ilinzweilin seemed to appreciate that setup better as well) and discuss propagating that change to the other tests we had already.

- [ ] We also need to standardise the error messages. Now some say "unknown-currency", others "unknown-token", etc

### CODE REVIEW

**Review by offerijns — COMMENTED** [review]
First pass, still diving into it more :)

File: src/Connector.sol [hunk 1]
```diff
@@ -238,6 +323,16 @@ contract CentrifugeConnector {
         memberlist.updateMember(user, validUntil);
     }
 
+    function handleTransfer(uint128 currency, address recipient, uint128 amount) public onlyGateway {
+        address currencyAddress = currencyIdToAddress[currency];
+        require(currencyAddress != address(0), "CentrifugeConnector/unknown-currency");
+
+        require(
```
[the remark this answers is not in the document]
Comment (offerijns) [reply src/Connector.sol:330]: Just leaving this here as a comment while I'm digging into why this assumption is wrong

[the remark this answers is not in the document]
Comment (offerijns) [reply src/Connector.sol:316]: Had to re-read this a few times, but I think the issue is as follows: The `escrow.approve` you are doing in the test is only possible because the escrow contract was deployed by the test contract, hence the test contract (which is the `msg.sender` of this `escrow.approve` call) is a `ward` on the escrow, see https://github.com/centrifuge/connectors/blob/94885c02662e4b6695f204e1409b473389a15aa3/src/Escrow.sol#L20. But in practice, no one else would be able to call this `escrow.approve`, except for the `Connector` contract, which will be a `ward` on the escrow.

So my suggestion is that within `Connector.handleTransfer`, this line should be added:
```
EscrowLike(escrow).approve(currencyAddress, address(this), amount);
```

File: src/Connector.sol [hunk 2]
```diff
@@ -129,71 +164,122 @@ contract CentrifugeConnector {
...
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
+
+        gateway.collectForInvest(poolId, trancheId, address(msg.sender), recipient);
     }
 
     // --- Incoming message handling ---
-    // todo(nuno): store currency and decimals
-    function addPool(uint64 poolId, uint128 currency, uint8 decimals) public onlyGateway {
+    function addCurrency(uint128 currency, address currencyAddress) public onlyGateway {
+        currencyIdToAddress[currency] = currencyAddress;
```
[the remark this answers is not in the document]
Comment (offerijns) [reply src/Connector.sol:236]: I think in the sync last week, we came to the conclusion that this should not be allowed for security reasons, and this should be immutable.

File: src/Connector.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol:317]: Thanks for looking into this! If `Escrow` is an immutable contract, how would this solution work if we deploy a new version of `Connector`? Would the old version of Connector need to add the new Connector address to the `wards` of the escrow?

I guess that when we deploy the Escrow, we first file the first Connector version as a ward and when/if we deploy a new version of Connector, we `deny` the former and `rely` on the latter?

[the remark this answers is not in the document]
Comment (offerijns) [reply src/Connector.sol:319]: I was just thinking about this as well :) In the deploy scripts, when deploying the escrow, the `msg.sender` will become a ward. Then, within that same deploy script, we should immediately `rely` some admin origin, and `deny` the `msg.sender`. So in the future, the admin origin can manage the wards on the escrow. And at some point we can add logic to Connectors for Centrifuge Chain governance to pass messages to manage this, and the admin origin becomes the backup.

File: src/Connector.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol:239]: Thanks, done. I think I found an edge case tho:

    // addCurrency(currency 1, address 0x123) -> Works
    // addCurrency(currency 1, address 0x456) -> Fails because there's already an address set for currency 1
    // addCurrency(currency 0, 0x123) -> Works because there's no address set for currency 0, even tho this same address is set for currency 1


In addition to `require(currencyIdToAddress[currency] == address(0),`, we should also check `require(currencyAddressToId[currencyAddress] == 0)`, to ensure that the same address doesn't get mapped to multiple currency ids. 

Or do we want to delegate this concern to Centrifuge? I.e, the mapping `Id <> EVM address` in Centrifuge should ensures that it's 1:1; This becomes an issue if we need to fix the `evm` address tho, since that requires creating a new mapping on Centrifuge 🤔

File: src/Connector.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol:321]: Perfect, that's a good setup. Will make these changes 👍

File: src/Connector.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (offerijns) [reply src/Connector.sol:241]: I think you're right that we should also add this check. What this would block, is ever re-using a currency address as a different ID; I'm trying to think through if there's any edge case where this might be required, but I can't think of anything.

> This becomes an issue if we need to fix the evm address tho, since that requires creating a new mapping on Centrifuge

I think this is already required by the first `require`, and will be the right path to fix this, yeah. Which means we should just do everything we can to avoid incorrect ID <> EVM address mappings :)

File: src/Connector.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol:322]: Made the changes the contract and the tests, left the deploy script changes to a separate issue: https://github.com/centrifuge/connectors/issues/62

File: src/Connector.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol:242]: Done ✅

File: src/Connector.sol [hunk 3]
```diff
@@ -10,16 +10,29 @@ interface GatewayLike {
...
+        external;
+    function decreaseInvestOrder(uint64 poolId, bytes16 trancheId, address investor, uint128 currency, uint128 amount)
+        external;
+    function increaseRedeemOrder(uint64 poolId, bytes16 trancheId, address investor, uint128 currency, uint128 amount)
+        external;
+    function decreaseRedeemOrder(uint64 poolId, bytes16 trancheId, address investor, uint128 currency, uint128 amount)
+        external;
+    function collectRedeem(uint64 poolId, bytes16 trancheId, address caller) external;
+    function collectInvest(uint64 poolId, bytes16 trancheId, address caller) external;
 }
```
Comment (offerijns) [inline src/Connector.sol:37]: I can't leave a comment on that line, but the `Pool` struct still contains a `currency` key that afaik can be removed.

File: src/Connector.sol [hunk 4]
```diff
@@ -126,75 +160,115 @@ contract CentrifugeConnector {
...
+
+    function addPool(uint64 poolId) public onlyGateway {
         Pool storage pool = pools[poolId];
         require(pool.createdAt == 0, "CentrifugeConnector/pool-already-added");
         pool.poolId = poolId;
         pool.createdAt = block.timestamp;
         emit PoolAdded(poolId);
     }
 
+    function allowPoolCurrency(uint128 currency, uint64 poolId) public onlyGateway {
```
Comment (offerijns) [inline src/Connector.sol:255]: This is definitely nit picking and I realize it would require changes on both sides, but, semantically wouldn't it make more sense if in the message and transaction parameters, `poolId` was before `currency`?

Similar to how if you are investing X in pool Y, then the arguments would be (pool Y, amount X).
And here you are allowing currency X for pool Y, so the arguments make sense to be (pool Y, currency X).

Comment (offerijns) [reply src/Connector.sol:255]: I also concluded this because the naming of the event was `PoolCurrencyAllowed`, as `CurrencyPoolAllowed` does not make sense.

**Review by offerijns — COMMENTED** [review]
Getting really close, mostly some more semantic comments now, but I think all the business logic is correct 👍

File: src/Connector.sol (hunk 4 above)
Comment (NunoAlexandre) [reply src/Connector.sol:255]: Yeah that makes sense, we should do that

File: src/Connector.sol [hunk 5]
```diff
@@ -10,16 +10,29 @@ interface GatewayLike {
...
+    function transfer(uint128 currency, address sender, bytes32 recipient, uint128 amount) external;
+    function increaseInvestOrder(uint64 poolId, bytes16 trancheId, address investor, uint128 currency, uint128 amount)
+        external;
+    function decreaseInvestOrder(uint64 poolId, bytes16 trancheId, address investor, uint128 currency, uint128 amount)
+        external;
+    function increaseRedeemOrder(uint64 poolId, bytes16 trancheId, address investor, uint128 currency, uint128 amount)
+        external;
+    function decreaseRedeemOrder(uint64 poolId, bytes16 trancheId, address investor, uint128 currency, uint128 amount)
+        external;
+    function collectRedeem(uint64 poolId, bytes16 trancheId, address caller) external;
```
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol:34]: Done!

File: src/Connector.sol [hunk 6]
```diff
@@ -40,15 +53,23 @@ struct Tranche {
...
 contract CentrifugeConnector {
     mapping(address => uint256) public wards;
     mapping(uint64 => Pool) public pools;
     mapping(uint64 => mapping(bytes16 => Tranche)) public tranches;
 
+    mapping(uint128 => address) public currencyIdToAddress;
+    // The reverse mapping of `currencyIdToAddress`
+    mapping(address => uint128) public currencyAddressToId;
+
+    // The currencies allowed for a given pool
```
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol:68]: Agreed, done

File: src/Connector.sol [hunk 7]
```diff
@@ -126,75 +160,115 @@ contract CentrifugeConnector {
...
+        require(token.hasMember(msg.sender), "CentrifugeConnector/not-a-member");
+
+        uint128 currency = currencyAddressToId[currencyAddress];
+        require(currency != 0, "CentrifugeConnector/unknown-currency");
+        require(poolCurrencies[poolId][currencyAddress], "CentrifugeConnector/pool-currency-not-allowed");
+
+        gateway.decreaseRedeemOrder(poolId, trancheId, msg.sender, currency, amount);
     }
 
     function collectRedeem(uint64 poolId, bytes16 trancheId) public {
```
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol]: Done and updated the spec ✅

File: src/Connector.sol [hunk 8]
```diff
@@ -126,75 +160,115 @@ contract CentrifugeConnector {
...
-    function collectForInvest(uint64 poolId, bytes16 trancheId, bytes32 userAddress) public {
-        // TODO(nuno)
+        gateway.collectInvest(poolId, trancheId, address(msg.sender));
     }
 
     // --- Incoming message handling ---
-    // todo(nuno): store currency and decimals
-    function addPool(uint64 poolId, uint128 currency, uint8 decimals) public onlyGateway {
+    function addCurrency(uint128 currency, address currencyAddress) public onlyGateway {
+        require(currencyIdToAddress[currency] == address(0), "CentrifugeConnector/currency-already-added");
```
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply src/Connector.sol:238]: Thanks for suggesting names :P done

File: src/Connector.sol (hunk 4 above)
Comment (NunoAlexandre) [reply src/Connector.sol:254]: Actually, the reason why I chose this order was that it reads "AllowPoolCurrency => allow currency C for pool P" instead of "allow, for pool P, currency C". But indeed for the investing messages we have poolId before so I am cool with keeping that format

File: src/Connector.sol (hunk 3 above)
Comment (NunoAlexandre) [reply src/Connector.sol:37]: Done, thanks for the catch

File: src/Connector.sol (hunk 4 above)
Comment (NunoAlexandre) [reply src/Connector.sol]: Done and updated the spec

**Review by offerijns — APPROVED** [review]
Looks ready to go to me 🚀 🔥
